fix(core): reset DI debug data once root view is removed

arturovt opened the pull request:
This commit resets the DI debug data when the root view is removed because its weak maps capture service instances that need to be garbage collected when the root view is removed. The debug data may be necessary only during app development but is not required when the app is destroyed.

---

Current behavior:

[Screencast from 01.02.24 22:30:04.webm](https://github.com/angular/angular/assets/7337691/6d4e9067-8556-4f53-a59e-669ad7532e24)

JoostK commented:
How would this work with multiple apps on the page, where only a subset is destroyed?

arturovt commented:
This would never work with multiple apps since, as far as I see, it's designed to be tied to a single app that currently exists. There's only one instance of the DIDebugData that always exists and is never re-assigned.

If it has to consider the case of multiple simultaneous apps on the page, it should then be a map of ApplicationRef -> DIDebugData. However, this would essentially require refactoring everything that already relies on the DIDebugData.

JoostK commented:
There may only be a single `DIDebugData`, but that is sufficient to support multiple apps. This change is the first to require registrations per app, so this would be the moment to change the data structure accordingly.

Since these are `WeakMap`s, I think the expectation would have been that none of this data can be a retainer, thus not contributing to memory leaks. Can you share more information on the memory leak you're observing and how this change avoids that problem?

arturovt commented:
Yeah, I literally meant that it has never supported the 'reset' functionality because its 'reset' implementation is tied to a single app. This would clean up everything once you have another app still alive, and the first one is being destroyed.

Initially, my understanding was that there should be no leak coming from the weak map side, as keys are instances of components and R3Injector classes. Let's take a closer look here:

![Screenshot from 2023-12-03 23-01-29](https://github.com/angular/angular/assets/7337691/940466f6-9489-4333-a151-566e81adeae9)

---

### And

And then when I reveal the summary view and find the final retainer:

![image](https://github.com/angular/angular/assets/7337691/daf2612a-9f29-4205-b7eb-20093f33ca4a)

I do understand what you mean and the expectation is obvious considering the weak map being used.

So, my question is, should we proceed with changing the implementation to a map of `new Map<ApplicationRef, DIDebugData>` that would create the debug data once the application ref is created and clean it up accordingly once it's destroyed?

JoostK commented:
I think `DIDebugData.reset` is mostly for testing purposes. Its only production usage is during initial setup.

That retainer is, ehm, interesting?! Does a forced garbage collect cycle make any difference?

Anyway, I suppose that keying this to an `ApplicationRef` would indeed be quite the undertaking, I don't know how feasible it even is. And in theory it shouldn't be necessary, I would like to understand why those `WeakMap`s would not avoid this problem.

/cc @AleksanderBodurri who has better understanding of the debugging infrastructure.

AleksanderBodurri commented:
@JoostK is correct about this function being used primarily in tests to reset state. `DIDebugData` is supported under the hood entirely by WeakMaps, and It seems unlikely to me that it would retain any memory there that would otherwise be garbage collected.

Looking at your images I notice the GENERATED_COMP_IDS constant, which seems to be a devmode only Map used to determine if component ids are colliding at runtime. If a component class declaration is retained in this map, it would not garbage collect, and thus not leave the DevTools WeakMap. I wonder if this could be causing the behaviour you're noticing

arturovt commented:
I'm testing applications separately to determine whether they're garbage-collectable in microfrontend mode. In this mode, applications are created and destroyed multiple times, and, in that particular case, they seem to resist garbage collection. Even though it's the development mode, I expect it to enhance development with additional assertions and helpful information. However, being in development mode shouldn't prevent items from being garbage-collected since these are unrelated matters.

The `GENERATED_COMP_IDS` retains the component constructor, which is always at the root context, but not the instance. The distance will always end up at the `GENERATED_COMP_IDS` root in snapshots because `lView` references constructor which is always retained in component IDs map. Using a weak map or a single map wouldn't matter in this case.

arturovt committed:
fix(core): reset DI debug data once root view is removed

This commit resets the DI debug data when the root view is removed
because its weak maps capture service instances that need to be garbage
collected when the root view is removed. The debug data may be necessary
only during app development but is not required when the app is destroyed.

AleksanderBodurri reviewed (commented):
Do you have a reproduction where I can see that these maps continue to retain references to LView and Injector instances that they shouldn't be? Perhaps I'm not understanding the problem exactly but these weak maps should not be holding any strong references at all
